tools/hdf_dump_translation_dataset.py

```diff
@@ -0,0 +1,325 @@
+    self.hdf_file.create_group("targets/labels")
+
+    self.hdf_file.create_dataset("seqLengths", (self.number_of_lines, 2), dtype="int32")
+    max_tag_length = len("line-") + len(str(self.number_of_lines))
+    self._tag_dtype = "S{}".format(max_tag_length)
+    self.hdf_file.create_dataset("seqTags", (self.number_of_lines,), dtype=self._tag_dtype)
+
+    self.hdf_file.create_dataset("inputs", (self.data_buffer_size,), maxshape=(None,),
+                                 dtype="int32", compression=self.compression_method)
+    self.hdf_file["targets/data"].create_dataset("classes", (self.data_buffer_size,), maxshape=(None,),
+                                                 dtype="int32", compression=self.compression_method)
+
+  def _write_attributes(self):
+    """
+    Writes several attributes to the HDF file.
+    """
+    self.hdf_file["targets/size"].attrs["classes"] = (self._vocabulary_sizes["target"], 1)
+
+    # Those should be deprecated, but include nevertheless to exactly reproduce hdf_dump.
+    self.hdf_file.attrs["inputPattSize"] = self._vocabulary_sizes["source"]
+    self.hdf_file.attrs["numLabels"] = self._vocabulary_sizes["target"]
+
+  def _write_labels(self):
+    """
+    Writes the labels (i.e. target vocabulary) to the HDF file.
+    """
+    sorted_vocabulary_tuples = sorted(self._vocabularies["target"].items(), key=itemgetter(1))
+    labels = [word.encode("utf8") for (word, _) in sorted_vocabulary_tuples]
+
+    assert len(labels) == self._vocabulary_sizes["target"], "Word ids were not unique."
+
+    max_label_length = max([len(label) for label in labels])
+    dtype = "S{}".format(max_label_length + 1)
+
+    labels = [numpy.array(label, dtype=dtype, ndmin=1) for label in labels]
+    labels = numpy.concatenate(labels)
+
+    self.hdf_file["targets/labels"].create_dataset("classes", (self._vocabulary_sizes["target"],),
+                                                   data=labels, dtype=dtype)
+
+  def _write_data(self):
+    """
+    Loads a chunk of lines from the corpus and writes all corresponding data to the HDF file.
+
+    :return: whether the end of the corpus is reached
+    :rtype: bool
+    """
+    source_chunk, target_chunk = self._get_chunk()
+
+    if not source_chunk:
+      self._finalize_data()
+      return True
+
+    source_lengths = [len(line) for line in source_chunk]
+    target_lengths = [len(line) for line in target_chunk]
+
+    chunk_size = len(source_chunk)
+
+    self._write_lengths(source_lengths, target_lengths)
+    self._write_tags(chunk_size)
+
+    self._write_data_indices(source_chunk, "source")
+    self._write_data_indices(target_chunk, "target")
+
+    self._number_of_processed_lines += chunk_size
+
+    return False
+
+  def _get_chunk(self):
+    """
+    Reads in the next chunk of lines from the corpus files.
+
+    :return: word indices for all lines in the current chunk, for both source and target
+    :rtype: tuple[list[numpy.ndarray[int]]]
```

Comment:
Better be more explicit about the type, i.e. `:rtype: list[numpy.ndarray], list[numpy.ndarray]`. I'm also not sure if `numpy.ndarray[int]` is valid. Did you check? I don't think it is valid.

Reply:
Also, the shape should be documented (in `:return:`) (just like the dtype).